pfsimulator/parflow_lib/problem_van_genuchten.c

```diff
@@ -0,0 +1,658 @@
+ *
+ *  Please read the COPYRIGHT file or Our Notice and the LICENSE file
+ *  for the GNU Lesser General Public License.
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License (as published
+ *  by the Free Software Foundation) version 2.1 dated February 1999.
+ *
+ *  This program is distributed in the hope that it will be useful, but
+ *  WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms
+ *  and conditions of the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public
+ *  License along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
+ *  USA
+ **********************************************************************EHEADER*/
+
+#include "parflow.h"
+
+/*--------------------------------------------------------------------------
+ * Structures
+ *--------------------------------------------------------------------------*/
+
+typedef struct {
+    int type;     /* input type */
+    void  *data;  /* pointer to Type structure */
+
+    NameArray regions;
+} PublicXtra;
+
+typedef struct {
+    Grid    *grid;
+
+    double  *temp_data;
+} InstanceXtra;
+
+typedef struct {
+  NameArray regions;
+  int num_regions;
+  int     *region_indices;
+  double  *values;
+} Type0;                       /* constant regions */
+
+typedef struct {
+    int num_regions;
+    int    *region_indices;
+    int data_from_file;
+    char   *alpha_file;
+    char   *n_file;
+    char   *s_sat_file;
+    char   *s_res_file;
+    double *alphas;
+    double *ns;
+    double *s_ress;
+    double *s_difs;
+    Vector *alpha_values;
+    Vector *n_values;
+    Vector *s_res_values;
+    Vector *s_sat_values;
+} Type1;                      /* Van Genuchten Saturation Curve */
+
+typedef struct {
+    int num_regions;
+    int    *region_indices;
+    double *alphas;
+    double *betas;
+    double *s_ress;
+    double *s_difs;
+} Type2;                      /* Haverkamp et.al. Saturation Curve */
+
+typedef struct {
+    int num_regions;
+    int    *region_indices;
+} Type3;                      /* Data points for Saturation Curve */
+
+typedef struct {
+    int num_regions;
+    int     *region_indices;
+    int     *degrees;
+    double **coefficients;
+} Type4;                      /* Polynomial function for Saturation Curve */
+
+typedef struct {
+    char    *filename;
+
+    Vector  *satRF;
+} Type5;                      /* Spatially varying field over entire domain
+                               * read from a file */
+
+/*--------------------------------------------------------------------------
+ * Saturation van genuchten write values into problemdata
+ *--------------------------------------------------------------------------*/
+void     vanGenuchten(
```

Comment:
New functions should use Doxygen style comments when merged to the main repo, see https://github.com/parflow/parflow/blob/master/CONTRIBUTING.md

For Doxygen type comment examples, see the comments in the new structs and functions in the file that is part of my recent PR: https://github.com/hokkanen/parflow/blob/CUDA11/pfsimulator/amps/mpi1/amps_gpupacking.cu